docs/proposals/010-new-diff-type-system.md

```diff
@@ -0,0 +1,283 @@
+
+See the next section for example uses of `[deriv_method]`.
+
+### 2. Split `IDifferentiable` interface: `IDifferentiableValueType` and `IDifferentiablePtrType`
+This approach moves away from "type-driven" derivative semantics and towards more "function-driven" derivative semantics.
+We no longer have a `dadd` , `dzero`, `dmul` etc.. we use default initialization instead of `dzero` and the backward derivative of the `use` method for `dadd`
+
+Further, `IDifferentiablePtrType` types don't have any of these properties. They do not need a way to 'add', and it is especially important that there is no default initializer. We never want the compiler to be able to create a new object of  `IDifferentiablePtrType` since we want to get the user-provided pointers.
+
+Additionally, we can use `IDifferentiableValueType` as the current `IDifferentiable` for backwards compatibility (it should just work in 95% of cases, since no one really defines dadd/dzero/dmul explicitly anyway) 
+
+Here's the new set of base interfaces:
+```csharp
+interface __IDifferentiableBase { } // Helper type for our implementation.
+interface IDifferentiableValueType : __IDifferentiableBase
+{
+	associatedtype Differential : IDifferentiableValueType & IDefaultInitializable;
+	[Differentiable] This use(); // auto-synthesized
+}
+
+interface IDifferentiablePtrType : __IDifferentiableBase
+{
+	associatedtype Differential : IDifferentiablePtrType;
+}
+
+```
+
+Some extras in stdlib allow us to constrain the diffpair type for things like `IArithmetic`
+```csharp
+// --- STDLIB EXTRAS ---
+
+interface ISelfDifferentiableValueType : IDifferentiableValueType
+{
+	// Force arithmetic types to be a differential pair of the same two types.
+	// Make it simple to define derivatives of arithmetic operations.
+	//
+	associatedtype Differential : This;
+}
+
+extension IFloat : ISelfDifferentiableValueType
+{ }
+
+extension float
+{
+	// trivial auto-synthesis (maybe we even prevent the user from overriding this)
+	float use() { return this; } 
+	
+	// trivial auto-synthesis (maybe we even prevent the user from overriding this).
+	[ForwardDerivativeOf(use)]
+	[deriv_method] void use_fwd() { return this; } 
+	
+	// auto-synthesized if necessary by invoking the use_bwd for all fields.
+	// we need to provide implementation for 'leaf' types.
+	[BackwardDerivativeOf(use)]
+	[deriv_method] [mutating] void use_bwd(float d) { this.d += d; } 
+}
+
+// The new system lets us define differentiable pointers easily.
+// IDifferentiablePtrType'd values are simply treated as references, so they can be freely
+// duplicated without requiring a `use()` for correctness.
+//
+struct DPtr<T : IDifferentiableValueType> : IDifferentiablePtrType
+{
+	typealias Differential = DPtr<T.Differential>;
+	
+	Buffer<T> buffer;
+	uint64 offset;
+	
+	[BackwardDerivative(get_bwd)]
+	[BackwardDerivative(get_fwd)]
+	T get() { return this.buffer[offset]; }
+	
+	[deriv_method] DifferentialPair<T> get_fwd()
```

Comment:
use_fwd?

Reply:
This is different from `use`. the `IDifferentiablePtrType` explicitly doesn't require a `use` (because we don't need any type-specific handling of bwd_diff(use)`. `use` is invoked internally by the compiler for any extra uses of an value-typed object (if used multiple times)

`get()` is a normal getter/setter method implemented on a buffer or pointer type (that a user can use) and is not required in the differentiable interface. 
It's effectively just derefencing or indexing in this scenario. (Maybe I should have used load() instead?). The core idea is that on these 'pointer-style' types all differentiation happens on the methods used to access the data, so we'll rely on custom derivatives, and simply make sure the pair is passed as-is through all the generated code so the buffer references are available to the custom derivative code.

Reply:
The main reason for introducing `IDifferentiablePtrType` is to make this explicit. It explicitly has no `use` and no way to initialize it, which means that this type will not get 'transposed'. They will behave similar to how `DiffTensorView<T>` works, with the main difference being that auto-diff itself will create the pair of buffers instead of us having to side-step the differentiable type system.